packages/gatsby/src/utils/page-data.ts

```diff
@@ -140,6 +141,28 @@ export async function flush(): Promise<void> {
     // them, a page might not exist anymore щ（ﾟДﾟщ）
     // This is why we need this check
     if (page) {
+      if (
+        program?._?.[0] === `develop` &&
+        process.env.GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND
+      ) {
+        // check if already did run query for this page
+        // with query-on-demand we might have pending page-data write due to
+        // changes in static queries assigned to page template, but we might not
+        // have query result for it
+        const query = queries.trackedQueries.get(page.path)
+        if (!query) {
+          // this should not happen ever
+          throw new Error(
+            `We have a page, but we don't have registered query for it (???)`
+          )
+        }
+
+        if (query.dirty !== 0) {
+          // query results are not up to date, it's not safe to write page-data and emit results
+          continue
+        }
+      }
+
```

Comment:
This is to not crash `gatsby develop` when there is pending data write because of static query list changing for a template (so for example when starting from cold cache, where we first assign initial list or when static query list change).

Note that this can put new `getPageData` util function in never resolving state for now (based on current implementation at least, we hope to handle this case - but function is only used when Query on Demand is enabled, so it won't impact users that don't opt into experimental feature)